<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [2]:
import pandas as pd
import psycopg2

In [3]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [4]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [5]:
# текст запроса
import warnings 
warnings.filterwarnings("ignore")
query_3_1 = f'''
select 
    count(*)
from vacancies
'''
df = pd.read_sql_query(query_3_1, connection)
df


count
0  49197

In [31]:
# результат запроса
# Количество вакансий 49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''
select 
    count(*)
from public.employers
'''
df = pd.read_sql_query(query_3_2, connection)
df

count
0  23501

In [33]:
# результат запроса
# Количество работадателей 23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [7]:
# текст запроса
query_3_3 = f'''
select 
    count(*)
from public.areas
'''
df = pd.read_sql_query(query_3_3, connection)
df

count
0   1362

In [35]:
# результат запроса
# Количество регионов 1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [8]:
# текст запроса
query_3_4 = f'''
select 
    count(*)
from public.industries
'''
df = pd.read_sql_query(query_3_4, connection)
df

count
0    294

In [37]:
# результат запроса
# Количество сфер деятельности 294

***

In [ ]:
# выводы по предварительному анализу данных
#Согласно имеющимся данным, в БД имеется 49147 вакансий от 23501 работодателя в 294 видах деятельности. Охватывает база 1362 региона.
connection.close()

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [9]:
# текст запроса
query_4_1 = f'''SELECT 
    a.name as area_name,
    COUNT(v.id) as vacansies_count
FROM VACANCIES v 
JOIN areas a ON v.area_id = a.id 
GROUP BY area_name
ORDER BY vacansies_count DESC
LIMIT 5
'''


In [10]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df


area_name  vacansies_count
0           Москва             5333
1  Санкт-Петербург             2851
2            Минск             2112
3      Новосибирск             2006
4           Алматы             1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [11]:
# текст запроса
query_4_2 = f'''SELECT 
    COUNT(v.id) as salary_count
FROM VACANCIES v 
WHERE (v.salary_from IS NOT NULL) OR (salary_to IS NOT NULL)

'''


In [12]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df
# Количество вакансий у которых заполнено хотя бы одно из двух полей зарплаты 24073

salary_count
0         24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [13]:
# текст запроса
query_4_3 = f'''SELECT 
    ROUND(AVG(v.salary_from)) as lower_salary,
    ROUND(AVG(v.salary_to)) as uper_salary
FROM VACANCIES v 
WHERE (v.salary_from IS NOT NULL) OR (salary_to IS NOT NULL)

'''
df = pd.read_sql_query(query_4_3, connection)
df

lower_salary  uper_salary
0       71065.0     110537.0

In [44]:
# результат запроса
# Средние значения для нижней границы зарплаты 71065.0, а для верхней границы 110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [14]:
# текст запроса
query_4_4 = f'''SELECT 
    v.schedule as "Тип рабочего графика",
    v.employment as "Тип трудоустройства",
    COUNT(id) as count_vacancies
FROM VACANCIES v 
GROUP BY v.schedule, v.employment
ORDER BY count_vacancies DESC
LIMIT 3

'''


In [15]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

Тип рабочего графика Тип трудоустройства  count_vacancies
0          Полный день    Полная занятость            35367
1     Удаленная работа    Полная занятость             7802
2        Гибкий график    Полная занятость             1593

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [16]:
# текст запроса
query_4_5 = f'''SELECT 
    v.experience as "Требуемый опыт работы",
    COUNT(id) as count_vacancies
FROM VACANCIES v 
GROUP BY v.experience
ORDER BY count_vacancies ASC
LIMIT 4

'''


In [17]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

Требуемый опыт работы  count_vacancies
0           Более 6 лет             1337
1             Нет опыта             7197
2         От 3 до 6 лет            14511
3    От 1 года до 3 лет            26152

***

In [ ]:
# выводы по детальному анализу вакансий
# Наибольшее количество вакансий в данной сфере наблюдается в столичных городах. Большая часть вакансий в размахом ЗП 71-110 тыс.руб. относится к числу вакансий с полным графиком.
# Большая часть предлагаемых вакансий предусматривает опыт работы от 1 до 3 лет. Менее 15% предназначено для лиц без опыта работы.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [18]:
# текст запроса
query_5_1 = f'''(SELECT 
    e.name as "Название работодателя",
    COUNT(v.id) as "Количество вакансий"
FROM public.employers e
JOIN public.vacancies v ON e.id = v.employer_id
GROUP BY e.name
ORDER BY COUNT(v.id) DESC
LIMIT 1)
UNION ALL
(
  SELECT 
    e.name as "Название работодателя",
    COUNT(v.id) as "Количество вакансий"
FROM public.employers e
JOIN public.vacancies v ON e.id = v.employer_id
GROUP BY e.name
ORDER BY COUNT(v.id) DESC
OFFSET 4
LIMIT 1  
)
'''


In [19]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

Название работодателя  Количество вакансий
0                Яндекс                 1933
1         Газпром нефть                  331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [20]:
# текст запроса
query_5_2 = f'''SELECT 
    a.name as "Название региона",
    COUNT(DISTINCT e.id) as "Количество работодателей",
    COUNT(DISTINCT v.id) as "Количество вакансий"
FROM public.areas a 
JOIN public.employers e ON e.area = a.id 
LEFT JOIN public.vacancies v ON v.area_id = a.id
WHERE v.id IS NULL
GROUP BY a.name
ORDER BY COUNT(e.id) DESC
LIMIT 1

'''


In [21]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

Название региона  Количество работодателей  Количество вакансий
0           Россия                       410                    0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [22]:
# текст запроса
query_5_3 = f'''SELECT 
    e.name as "Название работодателей",
    COUNT(DISTINCT v.area_id) as "Количество регионов где публикуются вакансии",
    COUNT(DISTINCT v.id) as "Количество вакансий"
FROM public.vacancies v 
JOIN public.employers e ON v.employer_id = e.id 
GROUP BY e.name
ORDER BY COUNT(DISTINCT v.id) DESC
LIMIT 10

'''


In [23]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

Название работодателей  Количество регионов где публикуют  \
0                 Яндекс                                181   
1             Ростелеком                                152   
2               Тинькофф                                 43   
3                   СБЕР                                 24   
4          Газпром нефть                                 24   
5            ИК СИБИНТЕК                                 46   
6                    МТС                                 55   
7                DataArt                                  2   
8  Совкомбанк Технологии                                 26   
9             Первый Бит                                 43   

   Количество вакансий  
0                 1933  
1                  491  
2                  444  
3                  428  
4                  331  
5                  327  
6                  292  
7                  247  
8                  204  
9                  176

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [24]:
# текст запроса
query_5_4 = f'''SELECT 
    COUNT(e.id) as "Количество работодателей"
FROM public.employers e 
LEFT JOIN public.employers_industries ie ON ie.employer_id = e.id
WHERE ie.industry_id IS NULL
'''

In [25]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

Количество работодателей
0                      8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [26]:
# текст запроса
query_5_5 = f'''SELECT 
    e.name as "Название работодателя",
    COUNT(DISTINCT ie.industry_id) as "Количество сфер деятельности"
FROM public.employers e 
LEFT JOIN public.employers_industries ie ON ie.employer_id = e.id
GROUP BY e.name
HAVING COUNT(DISTINCT ie.industry_id) = 4
OFFSET 2
LIMIT 1
'''


In [27]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

Название работодателя  Количество сфер деятельности
0                  2ГИС                             4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [28]:
# текст запроса
query_5_6 = f'''SELECT 
       COUNT(distinct e.id) as "Количество работодателей"
FROM  EMPLOYERS e   
LEFT JOIN EMPLOYERS_INDUSTRIES ie ON e.id = ie.employer_id
JOIN INDUSTRIES i ON i.id = ie.industry_id 
WHERE i.name = 'Разработка программного обеспечения'
'''

In [29]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

Количество работодателей
0                      3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [30]:
# код для получения списка городов-милионников
import pandas as pd
from bs4 import BeautifulSoup 
import requests 

url = "https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8"
response = requests.get(url, headers={'User-Agent': 'Mozilla/5.0'})
soup = BeautifulSoup(response.text, 'html.parser')
table = soup.find('table', class_="standard sortable")

if table:
    city_names = []
    rows = table.find_all("tr")[1:] 
    for row in rows:
        cols = row.find_all('td')
        if len(cols) > 1:
            city_name = cols[1].text.strip()
            city_names.append(city_name)
    for mln_sities in city_names:
        print(mln_sities)
    

Москва
Санкт-Петербург
Новосибирск
Екатеринбург
Казань
Красноярск
Нижний Новгород
Челябинск
Уфа
Краснодар
Самара
Ростов-на-Дону
Омск
Воронеж
Пермь
Волгоград


In [32]:
# текст запроса
query_5_7 = f''' (select 
    a.name as "name_area", 
    count(v.id) as "vacancy_count"
from 
    public.areas a
    left join public.vacancies v on v.area_id = a.id
    left join public.employers e on e.id = v.employer_id
where  e.name = 'Яндекс' 
    and a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Нижний Новгород', 'Казань', 'Челябинск', 'Самара', 'Омск', 'Ростов-на-Дону', 
    'Уфа', 'Красноярск', 'Пермь', 'Воронеж', 'Волгоград', 'Краснодар')
group by a.name
order by count(v.id) ASC)

union all

(select 
    'Total', 
    count(v.id)
from 
    public.areas a
    left join public.vacancies v on v.area_id = a.id
    left join public.employers e on e.id = v.employer_id
where  e.name = 'Яндекс' 
    and a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Нижний Новгород', 'Казань', 'Челябинск', 'Самара', 'Омск', 'Ростов-на-Дону', 
    'Уфа', 'Красноярск', 'Пермь', 'Воронеж', 'Волгоград', 'Краснодар'))
order by 2 ASC
'''



In [33]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

name_area  vacancy_count
0              Омск             21
1         Челябинск             22
2        Красноярск             23
3         Волгоград             24
4    Ростов-на-Дону             25
5             Пермь             25
6            Казань             25
7               Уфа             26
8            Самара             26
9         Краснодар             30
10          Воронеж             32
11      Новосибирск             35
12  Нижний Новгород             36
13     Екатеринбург             39
14  Санкт-Петербург             42
15           Москва             54
16            Total            485

***

In [ ]:
# выводы по анализу работодателей

### Выводы по анализу работодателей

Анализ работодателей показывает, что наибольшее количество вакансий компании находятся в Москве и СПб, однако их количество не столь велико в общей сумме. 
Города-миллионники азиатской части РФ за исключением Новосибирска - в числе отстающих. (Так как я проживаю в Красноярске, видимо, придется переезжать:). 
В целом, можно увидеть корреляцию между количеством вакансий и традиционными научными центрами страны.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [34]:
# текст запроса
query_6_1 = f''' 
SELECT
    COUNT(*) 
FROM public.vacancies v
WHERE v.name ILIKE '%data%' OR v.name ILIKE '%данн%'
''' 

In [35]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [36]:
# текст запроса
query_6_2 = f''' 
SELECT
    COUNT(*) 
FROM public.vacancies v
WHERE 
    (v.name ILIKE '%data scientist%' OR
        v.name ILIKE '%data science%'OR
        v.name ILIKE '%исследователь данных%' OR
        (v.name ILIKE '%ML%' AND v.name NOT ILIKE '%HTML%') OR
        v.name ILIKE '%machine learning%' OR v.name ILIKE '%машинн%обучен%')
        and
        (v.name ILIKE '%junior%' or
        v.experience = 'Нет опыта' or
        v.employment ILIKE '%стажировка%')
''' 

In [37]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [38]:
# текст запроса
query_6_3 = f'''
select count(*)
    from
        vacancies
    where
        (
        name ILIKE '%data scientist%'  or
        name ILIKE '%data science%' or
        name ILIKE '%исследователь данных%'  or
        (name ILIKE '%ML%' and name not ILIKE '%HTML%')  or
        name ILIKE '%machine learning%'  or
        name ILIKE '%машинн%обучен%'
        )
        and
        (
        key_skills ILIKE '%SQL%' or
        key_skills ILIKE '%postgres%'
        )
'''

In [39]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [40]:
# текст запроса
query_6_4 = f'''
select count(*)
    from
        vacancies
    where
        (
        name ILIKE '%data scientist%'  or
        name ILIKE '%data science%' or
        name ILIKE '%исследователь данных%'  or
        (name ILIKE '%ML%' and name not ILIKE '%HTML%')  or
        name ILIKE '%machine learning%'  or
        name ILIKE '%машинн%обучен%'
        )
        and
        (
        key_skills ILIKE '%python%' 
        )
'''

In [41]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [42]:
# текст запроса
query_6_5 = f'''
SELECT 
    ROUND(AVG(length(key_skills)-length(replace(key_skills, CHR(9), ''))+1), 2) as result
FROM VACANCIES v
WHERE 
(
  name ilike '%data scientist%'or
  name ilike '%data science%'or
  name ilike '%исследователь данных%'or
  (name like '%ML%' and name not ilike '%HTML%') or
  name ilike '%machine learning%' or name ilike '%машинн%обучен%'
)
'''

In [43]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

result
0    6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [44]:
# текст запроса
query_6_6 = f'''
SELECT 
  experience as "опыт работы",
    round(AVG(COALESCE((salary_from + salary_to) / 2, salary_from, salary_to)), 0) as "Средняя заработная плата"
FROM VACANCIES v
WHERE 
(
  name ilike '%data scientist%'or
  name ilike '%data science%'or
  name ilike '%исследователь данных%'or
  (name like '%ML%' and name not ilike '%HTML%') or
  name ilike '%machine learning%' or name ilike '%машинн%обучен%'
)
AND experience = 'От 3 до 6 лет'
AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
GROUP BY experience
'''

In [45]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

опыт работы  Средняя заработная плата
0  От 3 до 6 лет                  243115.0

***

## Выводы по предметному анализу

По данным анализа к сфере Анализа данных относится 1771 позиция (>3%). 
При этом, менее 3% из них предназначен для людей без опыта работы. 

Python  в целом имеет несколько большую популярность у работодателей, чем SQL. Обычно, указывается 6-7 навыков. 

Средняя зарплата Data-Scientist'а с опытом работы в данной базе составила 243115 руб./мес.

# Общий вывод по проекту

1. В целом проведенный анализ показал наличие вакансий в сфере Data-Science в России, однако количество их для начинающих специалистов было относительльно невелико. 
2. В целом, SQL является более подходящим инструментом для работы с БД, чем Excel, так как последний тратит слишком много времени для анализа таких объемов информации.
3. При дальнейшей практике в этой области, появляется возможность делать самые различные запросы по разным критериям.

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования